### DISCUSSION TITLE
Automatically skipping unmatched blocks when used as middleware

### DISCUSSION THREAD

**Discussion by janko** [jeremyevans/roda#334] [General]
In rodauth-rails, I've configured the Rodauth app to be used as a middleware. The problem is, people sometimes get tripped up about unmatched requests not reaching the Rails router. Let's use this example:

```rb
class Middleware < Roda
  plugin :middleware

  route do |r|
    r.on "auth" do
      r.post("login") { ... }
      r.post("create-account") { ... }
      # ...
    end
  end
end

class App
  # ... main app ...
end

use Middleware
run App
```

In the example above, any `/auth/*` requests not matched by terminal matchers will not automatically get forwarded to the main app. To address this, I've been recommending using the pass plugin:

```rb
plugin :pass

route do |r|
  r.on "auth" do
    r.post("login") { ... }
    r.post("create-account") { ... }
    # ...
    r.pass
  end
end
```

However, given that the Rodauth documentation cannot show this in every example, people get [tripped up](https://github.com/janko/rodauth-rails/discussions/231) by this.

Is there a way rodauth-rails could configure the Roda middleware to automatically exit from matchers that didn't handle the request? Would overriding `block_result_body` to call `pass` work, or is that too hacky?

```rb
module RequestMethods
  def block_result_body(result)
    case result
    when nil, false
      pass
    else
      super
    end
  end
end
```

What do you think about possibly introducing this functionality into the `middleware` plugin? My idea was to add a plugin option that calls `throw :next` instead of `pass` in the example above? I think that might even be more correct behavior, because it doesn't change how the routing tree works, just forwards unhandled requests.

Another idea was for rodauth-rails to use the `not_found` plugin and call `throw :next`. It does feel like it would be relying on internals of the middleware plugin a bit, though.

```rb
not_found do
  throw :next, true
end
```

**Comment by jeremyevans** [comment]
I think using the `not_found` approach is probably best for what you want.  I'm open to a middleware plugin option that sets that up automatically, so the `throw :next, true` internals don't have to leak out.

**Reply by jeremyevans** [reply]
Implemented in 05af2e52a5c54044bcb9d82d143ebe63845aae08

**Reply by janko** [reply]
Thank you! 🙏 I look forward to activating this once it's released 👍
